Cats instances taking the environmental parameter

gvolpe opened the pull request:
Now that `ZIO` became a trifunctor, interop instances should also take into account the environmental parameter `R` as well. Here's the summary of changes:

- A new type alias `type TaskR[-R, +A] = ZIO[R, Throwable, A]` was introduced.
- ~All existing tests except for one (`SemigroupK[UIO[?]]`) are passing. ~ I replaced it with both `Task` and `TaskR`. Does `UIO[?]` make any sense in this test?
- ~In some places where effects needs to be evaluated is necessary to convert `R` into `Any` via `asInstanceOf`, not sure if that's the best way.~ :white_check_mark: 

### TODO

- [x] Fix (or delete if correct) the commented test.
- [x] Add more tests summoning instances using different environmental parameters.

Consider this PR a proposal to further discuss the topic rather than a final solution :)

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,54 +30,56 @@ abstract class CatsPlatform extends CatsInstances {
 }
 
-private class CatsConcurrentEffect extends CatsConcurrent with effect.ConcurrentEffect[Task] {
+private class CatsConcurrentEffect[R] extends CatsConcurrent[R] with effect.ConcurrentEffect[TaskR[R, ?]] {
```
For this type class, I think we will have to hard code `R` to be some known set of environmental effects (say, all the effects supported by `DefaultRuntime`), so that we can provide them before we (for example) convert `toIO`.

Basically if you look at Cats `ReaderT[R, F, ?]`, we can support all the cats effects type classes that it supports, without hard coding `R`; then for the rest, we need to pick an `R`. Either that or use implicits to "inject" an `R`.

Let's have @kaishh take a look too since this could hugely improve usability of `R` for cats-effect users...

kaishh replied:
Yeah, saying `[R >: DefaultRuntime]` would allow an implementation of `toIO` - but it wouldn't be all that useful for someone trying to use `R` as Reader since they would be limited to only `R >: DefaultRuntime <: Any` - @gvolpe please correct me if you think otherwise. It would also invite incoherence, because the `DefaultRuntime` provided will NOT be the one created by user, but just `CatsConcurrentEffect`'s own `this`!

The most principled solution I see would unfortunately _decrease_ usability for cats-effect users :)
- Provide instances up to Concurrent for any `R`
- Provide an instance for ConcurrentEffect with Effect for `R <: Runtime[R]` where implicit R is required and STOP extending DefaultRuntime since we can use the executor from implicit instead of creating a global

Another is to simply provide *Effect only for `R=Any`, this means no implicit rituals and no incoherence in environment – at the cost of not being able to access CatsConcurrentEffect.this - but I don't think that's an awful loss.

gvolpe replied:
@jdegoes @kaishh thanks for looking into it. 

> Provide instances up to Concurrent for any R

I think this makes sense and this is the way it's done in `cats-effect`. Instances for `Kleisli` can only be implemented up to `Concurrent` but somehow I thought having the Reader capabilities built in would allow to go further.

> Another, more usable solution, is to simply provide *Effect only for R=Any

This seems to be the best solution given the first point :)

Will give it a spin later today.

jdegoes replied:
> Another, more usable solution, is to simply provide *Effect only for R=Any

Unless I'm missing something, I'm not a fan of this approach as it doesn't _increase_ usability.

I really think that for up to `Concurrent`, we should be maximally polymorphic and support _any_ `R`, because it has no impact on cats-effect users: if they're not using `R` (`R =:= Any`), then they can still benefit from the polymorphic instances; and if they _are_ using `R`, they'll be grateful that "just works" out of the box with everything up to `Concurrent` — and the only reason it doesn't work with classes beyond that is because of the run methods and `toIO` methods that will be going away in Cats Effect 2.0.

The "global injector" is interesting. Right now you have zero control over thread pool in Cats. You're stuck with the default one. No chance to use, for example, Scala's global execution context or an app-specific thread pool.

If we stop extending `DefaultRuntime`, then `Concurrent` and beyond could require an implicit `Runtime[R]`, which allows a user to do two things:

 - Use any `R` for all the type classes
 - Configure the platform (thread pool, fatal errors, error reporter)

This comes at cost to adding an implicit, but we could make it easy, e.g.:

```scala
import scalaz.zio.cats.implicits.DefaultRuntime
```

How does this sound?

gvolpe replied:
@jdegoes I think @kaishh meant `R=Any` for `Effect` and `ConcurrentEffect`. All the instances up to `Concurrent` are now fully polymorphic on `R`.

With regards to the runtime, I'll leave that decision to you both that are more familiar with the design :)

jdegoes replied:
All right, great. In that case, for `Concurrent` and beyond, my vote would be having those instances take an implicit `Runtime[R]` so we can make them all polymorphic in `R`, and then adding a global implicit so users can import to "unsafe run" stuff that just uses `R = Console with Random with System with Blocking with Scheduler`.

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -157,100 +162,110 @@ private class CatsEffect
   )(cb: Either[Throwable, A] => effect.IO[Unit]): effect.SyncIO[Unit] =
     effect.SyncIO {
-      this.unsafeRunAsync(fa) { exit =>
+      this.unsafeRunAsync(fa.asInstanceOf[Task[A]]) { exit =>
```
Ditto here. Basically anywhere you see `asInstanceOf`.

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -157,100 +162,110 @@ private class CatsEffect
+  // FIXME: asInstanceOf should not be needed
+  final override val parallel: ZIO[R, E, ?] ~> ParIO[E, ?] =
+    new (ZIO[R, E, ?] ~> ParIO[E, ?]) {
+      def apply[A](fa: ZIO[R, E, A]): ParIO[E, A] = Par(fa.provide(().asInstanceOf[R]))
```
Maybe we have to change the `ParIO` type synonym to make this one work without `asInstanceOf`?

gvolpe replied:
Definitely, I forgot about this one! :smile:

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,54 +30,56 @@ abstract class CatsPlatform extends CatsInstances {
     effect.SyncIO {
       this.unsafeRun {
-        fa.fork.flatMap { f =>
+        fa.asInstanceOf[Task[A]].fork.flatMap { f =>
```
Here

gvolpe commented:
@jdegoes thanks for reviewing. I mentioned the use of `asInstanceOf` in the description, I couldn't find another way to make it compile but given @kaishh 's input I think I can make it work without the hack. Allow me some time to get into it :)

gvolpe commented:
> Provide an instance for ConcurrentEffect with Effect for R <: Runtime[R] where implicit R is required and STOP extending DefaultRuntime since we can use the executor from implicit instead of creating a global

@kaishh I made the effect instances take an `implicit runtime: Runtime[Any]`, is that what you meant?

gvolpe commented:
The CI is complaining about formatting but I ran `scalafmt` and the tests locally without any issues.

```
[error] /home/travis/build/scalaz/scalaz-zio/interop-cats/jvm/src/test/scala/scalaz/zio/interop/catzSpec.scala isn't formatted properly!
[error] (interopCatsJVM / Test / scalafmtCheck) /home/travis/build/scalaz/scalaz-zio/interop-cats/jvm/src/test/scala/scalaz/zio/interop/catzSpec.scala isn't formatted properly!
[error] Total time: 5 s, completed Mar 4, 2019 5:29:13 AM
```

I've run `scalafmt` once again but there are no changes. Is there any other command I should run? Or maybe restarting the CI build could work?

ghostdogpr commented:
@gvolpe I restarted the `Lint` task but same result. Did you run `sbt fmt` from the root directory?

gvolpe commented:
Thanks @ghostdogpr , that did the trick. I was running just `scalafmt`.

kaishh commented:
@gvolpe
> I made the effect instances take an implicit runtime: Runtime[Any], is that what you meant?

No, the point is to allow injection of ANY `R`, which would be the behavior most useful to someone using Reader R, but as long as that R cake has a Runtime we can use to run the IO. One of `R <: Runtime[R]`, `R <: Runtime[_]`, `R <: Platform` or `R <:< Runtime[R]` should be able to express that.

gvolpe commented:
@kaishh do we really need to have `Runtime[R]` for the typeclasses up to `Concurrent`? I was thinking that in order to run your `TaskR[R, ?]` you need to `provide` the `R` first, in which case `Runtime[Any]` will do just fine.

Now if you think `Runtime[R]` makes sense is there any other way to provide it instead of mixing it in the cake? I'd try to avoid this at all cost because it wouldn't work with classy optics I think, which is what I'm aiming for (and got already working in a demo project).

How about `[R](implicit rts: Runtime[R])`? No mixing needed.

kaishh commented:
@gvolpe 
> do we really need to have Runtime[R] for the typeclasses up to Concurrent?

No, we don't. I'm only talking about ConcurrentEffect and Effect, we don't need to run and don't need the runtime for non *Effect classes.

> How about [R](implicit rts: Runtime[R])? No mixing needed.

Ok, I guess that makes the most sense if you're ok with it.

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,53 +30,72 @@ abstract class CatsPlatform extends CatsInstances {
+private abstract class CatsConcurrentEffect
+    extends CatsConcurrent[Any]
+    with effect.ConcurrentEffect[Task]
+    with effect.Effect[Task] {
```
Change these to `TaskR` for polymorphic `R`.

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,53 +30,72 @@ abstract class CatsPlatform extends CatsInstances {
+    extends CatsConcurrent[Any]
+    with effect.ConcurrentEffect[Task]
+    with effect.Effect[Task] {
+  def rts: Runtime[Any]
```
Change this to a constructor parameter: `Runtime[R]`, so you can `provide` the `R` before running or converting to Cats Effect `IO`.

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,53 +30,72 @@ abstract class CatsPlatform extends CatsInstances {
 
-  implicit val taskEffectInstances: effect.ConcurrentEffect[Task] with SemigroupK[Task] =
-    new CatsConcurrentEffect
+  implicit def taskEffectInstances(implicit runtime: Runtime[Any]): effect.ConcurrentEffect[Task] =
```
Make this polymorphic in `R` and change `runtime` to have type`Runtime[R]`.

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,53 +30,72 @@ abstract class CatsPlatform extends CatsInstances {
-  implicit val taskEffectInstances: effect.ConcurrentEffect[Task] with SemigroupK[Task] =
-    new CatsConcurrentEffect
+  implicit def taskEffectInstances(implicit runtime: Runtime[Any]): effect.ConcurrentEffect[Task] =
+    new CatsConcurrentEffect { override val rts = runtime }
```
Pass this in via constructor parameter.

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,53 +30,72 @@ abstract class CatsPlatform extends CatsInstances {
+    new CatsConcurrentEffect { override val rts = runtime }
 
-  implicit val taskParallelInstance: Parallel[Task, Util.Par] =
+  implicit def taskConcurrentInstances[R]: effect.Concurrent[TaskR[R, ?]] =
```
This instance will conflict with the above one, won't it? May have to use the "implicits priority trick"

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,53 +30,72 @@ abstract class CatsPlatform extends CatsInstances {
+  implicit def taskConcurrentInstances[R]: effect.Concurrent[TaskR[R, ?]] =
+    new CatsConcurrent[R]
+
+  implicit def taskParallelInstance(implicit rts: Runtime[Any]): Parallel[Task, Util.Par] =
```
Change to `Runtime[R]` and generalize `Task` to `TaskR`.

jdegoes replied:
Actually there's another parallel instance down below that will clash with this one, just delete this one.

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,53 +30,72 @@ abstract class CatsPlatform extends CatsInstances {
 
-private class CatsConcurrentEffect extends CatsConcurrent with effect.ConcurrentEffect[Task] {
+private abstract class CatsConcurrentEffect
+    extends CatsConcurrent[Any]
```
Change to `CatsConcurrent[R]`

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,53 +30,72 @@ abstract class CatsPlatform extends CatsInstances {
 }
 
-private class CatsConcurrentEffect extends CatsConcurrent with effect.ConcurrentEffect[Task] {
+private abstract class CatsConcurrentEffect
```
Introduce `R` type parameter.

ghostdogpr commented:
It would be nice to update the microsite as well with an example showing how to get an instance of ConcurrentEffect with this new way (the current doc doesn’t tell that you need an implicit clock).

gvolpe commented:
@jdegoes I believe I addressed all your comments. 

@ghostdogpr maybe the docs part can be another PR?

ghostdogpr commented:
> @ghostdogpr maybe the docs part can be another PR?

Sure!

gvolpe commented:
An unrelated test is failing on the CI build (no issues locally):

```
[error] Failed tests:
[error] 	scalaz.zio.stream.StreamSpec
```

ghostdogpr commented:
@gvolpe I relaunched and it passed 👍

jdegoes commented on interop-cats/jvm/src/test/scala/scalaz/zio/interop/catzSpec.scala:
```diff
@@ -102,25 +104,44 @@ class catzSpec
-        val v1  = unsafeRunSync(io1.timeout(20.seconds)).map(_.get)
-        val v2  = unsafeRunSync(io2.timeout(20.seconds)).map(_.get)
+      def eqv(io1: ZIO[R, E, A], io2: ZIO[R, E, A]): Boolean = {
+        val v1  = unsafeRunSync(io1.asInstanceOf[IO[E, A]].timeout(20.seconds)).map(_.get)
```
This needs to use `provide`

gvolpe replied:
How do I get the R in this case? By creating an `Arbitrary` for `R`?

jdegoes replied:
You need an implicit runtime. Or you can provide equality for `IO[E, A]` or even `Task[A]` if the tests will permit it. I don't think the tests use anything except for `Task`.

gvolpe replied:
@jdegoes as I mentioned on Gitter I'm still lost with this part. Adding an `implicit rts: Runtime[R]` constraint to derive the `Eq` instances makes sense but the issue is where to get this instance from? If you look at how the spec class is defined maybe you'll see the problem?

```scala
class catzSpec
    extends FunSuite
    with BeforeAndAfterAll
    with Matchers
    with Checkers
    with Discipline
    with TestInstances
    with GenIO
    with DefaultRuntime {

  override val Platform = PlatformLive.makeDefault().withReportFailure(_ => ())

  implicit val runtime: Runtime[Any] = this
```

If I remove the `with DefaultRuntime` how do I get a runtime? This class is instantiated by the test framework AFAIU.

ghostdogpr replied:
@gvolpe you can just do `implicit val runtime: Runtime[Any] = new DefaultRuntime {}`

gvolpe replied:
The question is how do I get a `Runtime[R]`? @ghostdogpr

ghostdogpr replied:
Sorry I missed your point. You're only using R = Int and R = String in the tests so isn't it sufficient to declare 2 implicit runtimes, one for Int and one for String with arbitrary values?

gvolpe replied:
I think that's not what John has in mind so I would love to know :D . We should be able to have tests for any kind of `R` and writing one for each doesn't seem convenient. I'm also very confused with this because it's exactly the reason why `cats-effect` doesn't provide instances of `ConcurrentEffect` and `Effect` for `Kleisli`. How do you run a function `A => F[B]`? What's the `A`?

jdegoes replied:
You need to get the R from somewhere. Define an implicit def that for any type `R: Arbitrary` can return a `Gen[Runtime[R]]`. Then you can easily summon a Runtime for any arbitrary R inside the test.

A more hacky way to do that is to define an implicit def that provides an inolicit Runtime for any Arbitrary R. By running the Gen to get some R.

gvolpe replied:
The arbitrary trick is what I suggested at first but we were probably thinking of different things :laughing: Now it makes sense, will try to get it done! Thanks :)

gvolpe replied:
@jdegoes Unfortunately I'm still stuck with this. The idea sounds good but it seems I require more than just `Runtime[R]`: I need `Runtime[R with Clock with System .... ]`. So this was my attempt:

```scala
  implicit val anyArb: Arbitrary[Any] = Arbitrary(Gen.const(()))

  type Env[R] = R with Clock with Console with System with Random with Blocking

  implicit def deriveRuntime[R: Arbitrary]: Gen[Runtime[Env[R]]] =
    Arbitrary.arbitrary[R].map { env =>
      Runtime[Env[R]](runtime.Environment + env, PlatformLive.Default)
    }

  implicit def catsEQ[R: Arbitrary, E, A: Eq](implicit rts: Runtime[Env[R]]): Eq[ZIO[R, E, A]] =
    new Eq[ZIO[R, E, A]] {
      import scalaz.zio.duration._

      def eqv(io1: ZIO[R, E, A], io2: ZIO[R, E, A]): Boolean = {
        val v1  = rts.unsafeRunSync(io1.timeout(20.seconds)).map(_.get)
        val v2  = rts.unsafeRunSync(io2.timeout(20.seconds)).map(_.get)
        val res = v1 === v2
        if (!res) {
          println(s"Mismatch: $v1 != $v2")
        }
        res
      }
    }

  implicit def catsParEQ[R: Arbitrary, E: Eq, A: Eq](implicit rts: Runtime[Env[R]]): Eq[ParIO[R, E, A]] =
    new Eq[ParIO[R, E, A]] {
      def eqv(io1: ParIO[R, E, A], io2: ParIO[R, E, A]): Boolean =
        rts.unsafeRun(Par.unwrap(io1).either) === rts.unsafeRun(Par.unwrap(io2).either)
    }
```

In order to create the environment I need to mixin some traits (`Runtime.apply`) and all I've got is a plain value of type `R` so this doesn't seem possible.

Also notice that where I require an `implicit rts: Runtime[Env[R]]` I'm not going to get one. Instead I'll have a `Gen` in scope to create a `Runtime[Env[R]]` which seems overkill. Why not just casting the given `ZIO[R, E, A]` to `IO[E, A]` as I've done to get it working? After all it's just tests to `unsafeRunSync` some computation but if there's a better way I'd like to know it.

So once again I'm on a dead end here :disappointed:

jdegoes replied:
Any attempt to cast away the `R` will result in a runtime exception somewhere, unless the `R` is not actually used. 

I see now that you cannot use an arbitrary `R` here. If you like, you could create a "test" module, like this one:

```scala
trait TestModule[+A] {
  def testModule: A
}
```

Then you could make an `Arbitrary` for that (given an `Arbitrary` for `A`), and finally, you could supply an `Arbitrary[Runtime[TestModule[A] with Clock with System ...]`.

However, the question is why? In my opinion, testing with `R = Int` and `R = String` like is being done is not that useful. You may as well fix the environment to `Clock with System` or weaker in the tests, rather than using `R = Int` or `R = String`.

If you fix the environment to the above, you can just use `implicit val TestRuntime = new DefaultRuntime`. Then fix the two or so compiler errors where a test is using a different `R`. Then you should be good to go.

It's a nice idea in theory to make the test `R agnostic` but you can't use a single `R` for the whole test suite in that case, which will cause lots of pain. Just best to fix the `R` for the test suite, because it's already as polymorphic as it can be in the instances.

gvolpe replied:
Thanks for bearing with me, John.

> However, the question is why? In my opinion, testing with R = Int and R = String like is being done is not that useful. You may as well fix the environment to Clock with System or weaker in the tests, rather than using R = Int or R = String.

You're right. I just want to verify that there are Cats instances in scope for `TaskR[R, ?]` so I could just add the code that summons such instances somewhere, don't need property tests. If something is wrong, the code just won't compile.

> If you fix the environment to the above, you can just use implicit val TestRuntime = new DefaultRuntime. Then fix the two or so compiler errors where a test is using a different R. Then you should be good to go.

I don't understand this. Having a `DefaultRuntime` works fine for the other tests but what do you mean with "fix the compiler errors where a test is using a different `R`"? I think this is the whole issue and why I'm stuck here, it needs a `Runtime[R]` :smile: 

> It's a nice idea in theory to make the test R agnostic but you can't use a single R for the whole test suite in that case, which will cause lots of pain. Just best to fix the R for the test suite, because it's already as polymorphic as it can be in the instances.

Completely agree :+1: I just want to make this as simple as possible, I just haven't found the way.

jdegoes commented on interop-cats/jvm/src/test/scala/scalaz/zio/interop/catzSpec.scala:
```diff
@@ -102,25 +104,44 @@ class catzSpec
-        val v2  = unsafeRunSync(io2.timeout(20.seconds)).map(_.get)
+      def eqv(io1: ZIO[R, E, A], io2: ZIO[R, E, A]): Boolean = {
+        val v1  = unsafeRunSync(io1.asInstanceOf[IO[E, A]].timeout(20.seconds)).map(_.get)
+        val v2  = unsafeRunSync(io2.asInstanceOf[IO[E, A]].timeout(20.seconds)).map(_.get)
```
This needs to use `provide`.

jdegoes commented on interop-cats/jvm/src/test/scala/scalaz/zio/interop/catzSpec.scala:
```diff
@@ -129,18 +150,23 @@ class catzSpec
+  implicit def catsParEQ[R, E: Eq, A: Eq]: Eq[ParIO[R, E, A]] =
+    new Eq[ParIO[R, E, A]] {
+      def eqv(io1: ParIO[R, E, A], io2: ParIO[R, E, A]): Boolean =
+        unsafeRun(Par.unwrap(io1.asInstanceOf[ParIO[Any, E, A]]).either) === unsafeRun(
```
This needs to use `provide`

jdegoes commented on interop-cats/jvm/src/test/scala/scalaz/zio/interop/catzSpec.scala:
```diff
@@ -129,18 +150,23 @@ class catzSpec
+    new Eq[ParIO[R, E, A]] {
+      def eqv(io1: ParIO[R, E, A], io2: ParIO[R, E, A]): Boolean =
+        unsafeRun(Par.unwrap(io1.asInstanceOf[ParIO[Any, E, A]]).either) === unsafeRun(
+          Par.unwrap(io2.asInstanceOf[ParIO[Any, E, A]]).either
```
This needs to use `provide`

kaishh commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,53 +30,69 @@ abstract class CatsPlatform extends CatsInstances {
-  implicit def ioTimer[R <: Clock, E]: effect.Timer[ZIO[R, E, ?]] = new effect.Timer[ZIO[R, E, ?]] {
-    override def clock: cats.effect.Clock[ZIO[R, E, ?]] = new effect.Clock[ZIO[R, E, ?]] {
-      override def monotonic(unit: TimeUnit): ZIO[R, E, Long] =
+  implicit def ioTimer[R <: Clock]: effect.Timer[TaskR[R, ?]] = new effect.Timer[TaskR[R, ?]] {
```
ioTimer shouldn't be restricted to Task, should be available for ZIO[R, E, A]

kaishh commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,53 +30,69 @@ abstract class CatsPlatform extends CatsInstances {
 
-  implicit val taskEffectInstances: effect.ConcurrentEffect[Task] with SemigroupK[Task] =
-    new CatsConcurrentEffect
+  implicit def taskEffectInstances[R](implicit runtime: Runtime[R]): effect.ConcurrentEffect[TaskR[R, ?]] =
```
`with SemigroupK[TaskR[R, ?]]` since that's available

kaishh commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,53 +30,69 @@ abstract class CatsPlatform extends CatsInstances {
 
-  implicit def parallelInstance[E](implicit M: Monad[IO[E, ?]]): Parallel[IO[E, ?], ParIO[E, ?]] =
-    new CatsParallel[E](M)
+  implicit def taskConcurrentInstances[R]: effect.Concurrent[TaskR[R, ?]] =
```
`with SemigroupK[Task]` since that's available

gvolpe replied:
Not `TaskR[R, ?]` here?

kaishh replied:
Sorry, fixed

gvolpe commented:
@jdegoes Also what do you think about renaming `TaskR` to `RIO`? It'll be like the Haskell implementation.

gvolpe commented on core/shared/src/main/scala/scalaz/zio/package.scala:
```diff
@@ -22,9 +22,10 @@ package object zio extends EitherCompat {
-  type UIO[+A]    = ZIO[Any, Nothing, A]
+  type IO[+E, +A]    = ZIO[Any, E, A]
+  type Task[+A]      = ZIO[Any, Throwable, A]
+  type TaskR[-R, +A] = ZIO[R, Throwable, A]
```
@jdegoes Something like this?

```scala
object TaskR extends ZIO_E_Throwable {
  type UpperE = Throwable

  def apply[R, A](a: => A): TaskR[R, A] = effect(a)
}
```

Or this?

```scala
object TaskR extends ZIO_E_Throwable {
  type UpperE = Throwable

  def apply[A](a: => A): Task[A] = effect(a)
}
```

I'm confused by the `IO` object that returns a `Task`.

jdegoes replied:
An `IO` object can return a task because a task can always be interpreted as an `IO` for some choice of `E` and `A`. The reverse is not true.

So the latter one is correct because a `Task` can always be interpreted as a `Task`. The former one is also correct but contains a useless `R` parameter that will only cause type inference problems (Scala can't infer it).

jdegoes commented on interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala:
```diff
@@ -30,11 +30,11 @@ abstract class CatsPlatform extends CatsInstances {
-      IO.yieldNow
+  implicit def ioContextShift[R]: ContextShift[TaskR[R, ?]] = new ContextShift[TaskR[R, ?]] {
+    override def shift: TaskR[R, Unit] =
+      ZIO.yieldNow
```
Once you add the companion object, you can change all these methods to use `TaskR` companion object.

kaishh replied:
`ContextShift` should be available for `ZIO[R, E, A]`

gvolpe commented:
@jdegoes I just removed all the property tests previously added with the new parameter `R` and left the `Eq` instances as they were more or less. 

Instead I added some code that summons the Cats instances that have an environmental parameter. This should do I hope.

jdegoes commented:
![giphy](https://user-images.githubusercontent.com/156745/54068181-1fcc3300-424a-11e9-8348-c2ea1f9cd5cc.gif)

Congratulations, & thanks for your contribution!
